cmd/agent/agent.go

```diff
@@ -47,105 +42,113 @@ var (
 	commit string
 )
 
-// AgentService represents the main structure for managing cloud executors and configuration.
-// It also embeds the gRPC server interface for handling gRPC requests.
-type AgentService struct {
-	cfg       *config.AgentConfig
-	executors map[string]cexec.CloudExecutor
-	logger    *zap.Logger
-	pb.UnimplementedAgentServiceServer
-}
-
 // init initializes the global logger configuration.
 // This is automatically invoked before the main function.
 func init() {
 	// Initialize logging configuration
 	logger = ciqLogger.NewLogger()
 }
 
-// NewAgentService creates and initializes a new AgentService instance.
-//
-// Parameters:
-//   - cfg: Pointer to AgentConfig containing the configuration details.
-//   - logger: Pointer to zap.Logger for logging.
-//
-// Returns:
-//   - *AgentService: A pointer to the newly created Agent instance.
-func NewAgentService(cfg *config.AgentConfig, logger *zap.Logger) *AgentService {
-	return &AgentService{
-		cfg:       cfg,
-		executors: make(map[string]cexec.CloudExecutor, 0),
-		logger:    logger,
-	}
+// Agent represents the main structure for the ClusterIQ Agent. It includes the
+// AgentService for implementing the gRPC server and the AgentCron for getting
+// and executing the Scheduled actions to be run by the AgentService when it's
+// needed
+type Agent struct {
+	cfg            *config.AgentConfig
+	ias            *InstantAgentService
+	cas            *ScheduleAgentService
+	eas            *ExecutorAgentService
+	actionsChannel chan actions.Action
+	logger         *zap.Logger
+	wg             *sync.WaitGroup
 }
 
-// AddExecutor adds a new CloudExecutor to the AgentService.
-//
-// Parameters:
-//   - exec: CloudExecutor instance to add.
-//
-// Returns:
-//   - error: An error if the executor is nil; otherwise, nil.
-func (a *AgentService) AddExecutor(exec cexec.CloudExecutor) error {
-	if exec == nil {
-		return fmt.Errorf("Cannot add a nil Executor")
-	}
+func NewAgent(cfg *config.AgentConfig, logger *zap.Logger) *Agent {
+	var ch chan actions.Action
+	var wg sync.WaitGroup
 
-	a.executors[exec.GetAccountName()] = exec
-
-	return nil
-}
+	ch = make(chan actions.Action)
 
-// readCloudProviderAccounts reads cloud provider account configurations from the credentials file.
-//
-// Returns:
-//   - []credentials.AccountConfig: A slice of account configurations.
-//   - error: An error if reading the file fails.
-func (a *AgentService) readCloudProviderAccounts() ([]credentials.AccountConfig, error) {
-	accounts, err := credentials.ReadCloudAccounts(a.cfg.Credentials.CredentialsFile)
-	if err != nil {
-		return nil, err
+	// Creating InstantAgentService (gRPC)
+	ias := NewInstantAgentService(&cfg.InstantAgentServiceConfig, ch, &wg, logger)
+	if ias == nil {
+		logger.Error("Cannot create InstantAgentService")
```

Comment:
should we exit here? if the agent doesn't start successfully, do we lose part/full functionality?

Reply:
Added error to return vars and handling on top-level function